### PR TITLE
[Improvement-13476][*]  Improve DS load hdfs configuration automatically by using environment variable HADOOP_CONF_DIR

### LINKED ISSUE #13476: [Improvement][*] Improve DS load hdfs configuration automatically by using environment variable `HADOOP_CONF_DIR`
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

In a complex environment, such as NameNode HA, authentication with Kerberos, and so on, DS needs more HDFS configuration from `hdfs-site.xml` and `core-site.xml` instead of only HDFS URL (`fs.defaultFS`). 
So user have to copy `hdfs-site.xml` and `core-site.xml` to the classpath manually;
 

### What you expected to happen

When the user config the environment `HADOOP_CONF_DIR`, DS can load HDFS configuration automatically  from envirement;

### How to reproduce

*

### Anything else

_No response_

### Version

dev

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by xxjingcd**
## Purpose of the pull request

Improve DS load hdfs configuration automatically by using common environment variables `HADOOP_CONF_DIR`  or `HADOOP_HOME` which are  usually  already configured in os;

It provides a more convenient choice. If the environment variable does not exist or is an error configuration, DS will load the configuration  as before;

This closes #13476

## Brief change log

- `HdfsStorageOperator`
- `dolphinscheduler_env.sh` : add tips for  environment variable`HADOOP_HOME` and `HADOOP_CONF_DIR`

## Verify this pull request

This change added tests and can be verified as follows:

- `HdfsStorageOperatorTest`

### CODE REVIEW

**Comment by ruanwenjun** [comment]
Please use `mvn spotless:apply` to fix the code style.

**Review by Radeity — COMMENTED** [review]
I think we should better avoid using system env, maybe we can add `hadoop.conf.dir` in our property file. 

In addition, i think there's nothing wrong with current way which move configuration files into classpath, i just wonder do we really have to make this change?

**Comment by zhongjiajie** [comment]
please send to detail to `private@dolphinscheduler.apache.org` @xxjingcd

**Comment by xxjingcd** [comment]
> In addition, i think there's nothing wrong with current way which move configuration files into classpath, i just wonder do we really have to make this change?

It is not  convenient enough. You have to copy hdfs-site.xml and core-site.xml to the classpath manually; And it will bring maintenance costs,  users  have to remember that  copy again when configuration files change ;

>I think we should better avoid using system env, maybe we can add hadoop.conf.dir in our property file.

It is a choice for users , you can still move configuration files into classpath. This only takes effect if you specify an environment variable；
**And the environment variables are usually already configured in the operating system;  That mean users can load correct configuration without doing anything.** 
It is a general approach;

**Comment by xxjingcd** [comment]
> Please use `mvn spotless:apply` to fix the code style.

I've already fixed it

**Comment by xxjingcd** [comment]
> I think we should better avoid using system env, maybe we can add `hadoop.conf.dir` in our property file.
> 
> In addition, i think there's nothing wrong with current way which move configuration files into classpath, i just wonder do we really have to make this change?

 Environment variables  `HADOOP_CONF_DIR` and `HADOOP_HOME` is very common in most components, e.g HBASE, Spark、Flink, and so on.

**Comment by Radeity** [comment]
> Environment variables HADOOP_CONF_DIR and HADOOP_HOME is very common in most components, e.g HBASE, Spark、Flink, and so on.

You are right, `HADOOP_CONF_DIR` is a common environment variable, **HOWEVER**, i think add it into properties is more general, especially when deploying in k8s.

**Comment by xxjingcd** [comment]
> please send to detail to `private@dolphinscheduler.apache.org` @xxjingcd

what detail？ Is it a mistake?

**Comment by xxjingcd** [comment]
> > Environment variables HADOOP_CONF_DIR and HADOOP_HOME is very common in most components, e.g HBASE, Spark、Flink, and so on.
> 
> You are right, `HADOOP_CONF_DIR` is a common environment variable, **HOWEVER**, i think add it into properties is more general, especially when deploying in k8s.

In practice, users always maintain a set of common global variables， this always include `HADOOP_CONF_DIR` and `HADOOP_HOME`.
 In a word, **using environment variables will make DS deploy faster and easily**. That is the reason why this pr use environment variables;

Actually, all I know about components are use environment variables to configure. For Example, a source code fragment of  Flink’s [HadoopUtils](https://github.com/apache/flink/blob/master/flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/util/HadoopUtils.java) 
 
```java 
    public static String[] possibleHadoopConfPaths(
            org.apache.flink.configuration.Configuration flinkConfiguration) {
        String[] possiblePaths = new String[4];
        possiblePaths[0] = flinkConfiguration.getString(ConfigConstants.PATH_HADOOP_CONFIG, null);
        possiblePaths[1] = System.getenv("HADOOP_CONF_DIR");

        if (System.getenv("HADOOP_HOME") != null) {
            possiblePaths[2] = System.getenv("HADOOP_HOME") + "/conf";
            possiblePaths[3] = System.getenv("HADOOP_HOME") + "/etc/hadoop"; // hadoop 2.2
        }
        return Arrays.stream(possiblePaths).filter(Objects::nonNull).toArray(String[]::new);
    }
```



At Last, we should not put all configurations in properties， especially if the configuration is contained in common environment variables. This will make DS deploy fast.
DS has a env config file `dolphinscheduler_env.sh`, similar to `spark-env.sh`. And if users want to overwrite global variables, just config `dolphinscheduler_env.sh` file. 

There is not much difference between putting it in `dolphinscheduler_env.sh`  and putting it in properties;

I know very little about k8s.  Maybe you can use `env` in dockerfile or config `dolphinscheduler_env.sh` similar to how you configure properties files.

**Comment by EricGao888** [comment]
> > please send to detail to `private@dolphinscheduler.apache.org` @xxjingcd
> 
> what detail？ Is it a mistake?

@xxjingcd I think the `detail` stuff @zhongjiajie comments here may refer to #13460

**Comment by Radeity** [comment]
> I know very little about k8s. Maybe you can use env in dockerfile or config dolphinscheduler_env.sh similar to how you configure properties files.

@xxjingcd Sorry for my mistake, we have already created environment variables with `configMap` in k8s deploy mode, including configuration `HADOOP_CONF_DIR`. 

> Is this convenient for users to configure HDFS? 

In this PR, i think you don't have to change that part of codes, it's up to users whether override hdfs configuration or not.

> Whether it is convenient enough to configure environment variables `HADOOP_CONF_DIR`.

For me, there's no problem, please update docs in resource center configuration page : )

Also, you don't have to add environment configuration in `dolphinscheduler_env.sh` (refer to https://github.com/apache/dolphinscheduler/pull/13028).

Basically LGTM, cc @zhongjiajie @EricGao888

**Comment by xxjingcd** [comment]
@Radeity thanks for your code review work. 

>In this PR, i think you don't have to change that part of codes, it's up to users whether override hdfs configuration or not.

OK, I'll remove the `todo` (about the question). 


> Also, you don't have to add environment configuration in dolphinscheduler_env.sh (refer to https://github.com/apache/dolphinscheduler/pull/13028).

The purpose of the pr #13028 is to remove the export of `dolphinscheduler_env.sh`  in the task’s initialization, so it deleted the environment configuration used in user's task. But it make a mistake, `JAVA_HOME` should not be deleted, because  `JAVA_HOME` used in  `dolphinscheduler-dameon.sh`.
After this pr, `dolphinscheduler_env.sh` is only used by DS component (used in `dolphinscheduler-dameon.sh` ) and DS interval, not by the user’s task; And after pr #13317, the user's tasks can specify the env file config by **`shell.env_source_list`**; **That means environment file used in DS and user's task is separated now**. And I have run some tasks to verify it.
So I think it's no problem to do that. Or maybe I don't get what you mean?

>For me, there's no problem, please update docs in resource center configuration page :

OK, I'll sync the configuration.

**Comment by Radeity** [comment]
> So I think it's no problem to do that. Or maybe I don't get what you mean?

We should better only add ds' internal env configuration, which we don't have now. And user/tenant manages external env configurations by themselves. `$JAVA_HOME` belongs to external configurations, we just have to declare them in the doc. That's my personal understanding of that pr, have to further make sure. cc @ruanwenjun

**Comment by xxjingcd** [comment]
> > So I think it's no problem to do that. Or maybe I don't get what you mean?
> 
> We should better only add ds' internal env configuration, which we don't have now. And user/tenant manages external env configurations by themselves. `$JAVA_HOME` belongs to external configurations, we just have to declare them in the doc. That's my personal understanding of that pr, have to further make sure. cc @ruanwenjun

Ok, I will remove `$JAVA_HOME` to make this pr‘s purpose simple.
